src/Parsers/ASTSelectQuery.cpp

```diff
@@ -438,4 +438,19 @@ ASTPtr & ASTSelectQuery::getExpression(Expression expr)
     return children[positions[expr]];
 }
 
+void ASTSelectQuery::setFinal()
+{
+    auto & tables_in_select_query = tables()->as<ASTTablesInSelectQuery &>();
+
+    if (tables_in_select_query.children.empty())
+        throw Exception(ErrorCodes::LOGICAL_ERROR, "Tables list is empty, it's a bug");
+
+    auto & tables_element = tables_in_select_query.children[0]->as<ASTTablesInSelectQueryElement &>();
```

Comment:
But is it OK to set FINAL only for one table?

Reply:
Yes, it's checked for the first table only:
https://github.com/ClickHouse/ClickHouse/blob/fd914f3d9715df06bef5611591b7852b8a9dc66c/src/Parsers/ASTSelectQuery.cpp#L299-L306